File: backend/opportunity/watchlist.go
package opportunity

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	"go-stock/backend/data"
	"go-stock/backend/db"
	"go-stock/backend/models"
	"go-stock/backend/papertrading"
)

// WatchlistItem is a read-only row for GET /api/watchlist (current WATCHING only).
type WatchlistItem struct {
	OpportunityID    string    `json:"opportunity_id,omitempty"`
	StockCode        string    `json:"stock_code"`
	StockName        string    `json:"stock_name,omitempty"`
	ScanBatchKey     string    `json:"scan_batch_key"`
	Source           string    `json:"source,omitempty"`
	WatchTime        time.Time `json:"watch_time"`
	LatestAction     string    `json:"latest_action"`
	InTradePlan      bool      `json:"in_trade_plan"`
	TradePlanStatus  string    `json:"trade_plan_status"`
	// TradePlanID is the representative plan id for this stock_code (0 when none).
	TradePlanID uint `json:"trade_plan_id,omitempty"`
}

// WatchlistView is GET /api/watchlist response body.
type WatchlistView struct {
	AccountID      uint            `json:"account_id"`
	WatchingCount  int             `json:"watching_count"`
	Items          []WatchlistItem `json:"items"`
	DataSourceNote string          `json:"data_source_note,omitempty"`
}

// ListWatchingQuery parameters for account-level watching list.
type ListWatchingQuery struct {
	AccountID uint
	Limit     int
}

// ListWatching returns opportunities whose latest user action is WATCH (append-only latest wins).
// IGNORE (and any non-WATCH latest) are excluded. No schema changes.
func ListWatching(q ListWatchingQuery) (*WatchlistView, error) {
	if err := EnsureSchema(db.Dao); err != nil {
		return nil, err
	}
	accountID := q.AccountID
	if accountID == 0 {
		acc, err := papertrading.GetDefaultAccount()
		if err != nil {
			return nil, err
		}
		accountID = acc.ID
	}
	limit := q.Limit
	if limit <= 0 {
		limit = 100
	}
	if limit > 500 {
		limit = 500
	}

	var rows []UserOpportunityAction
	err := db.Dao.Where("account_id = ?", accountID).
		Order("created_at DESC").
		Find(&rows).Error
	if err != nil {
		return nil, err
	}

	seen := map[string]struct{}{}
	items := make([]WatchlistItem, 0)
	for i := range rows {
		oid := strings.TrimSpace(rows[i].OpportunityID)
		if oid == "" {
			continue
		}
		if _, ok := seen[oid]; ok {
			continue
		}
		seen[oid] = struct{}{}
		if rows[i].Action != ActionWatch {
			continue
		}
		name, source := enrichWatchlistMeta(rows[i].ScanBatchKey, rows[i].StockCode)
		items = append(items, WatchlistItem{
			OpportunityID:   oid,
			StockCode:       rows[i].StockCode,
			StockName:       name,
			ScanBatchKey:    rows[i].ScanBatchKey,
			Source:          source,
			WatchTime:       rows[i].CreatedAt,
			LatestAction:    ActionWatch,
			InTradePlan:     false,
			TradePlanStatus: WatchlistTradePlanNone,
		})
	}

	watchingCount := len(items)
	if len(items) > limit {
		items = items[:limit]
	}
	enrichWatchlistTradePlan(items)

	return &WatchlistView{
		AccountID:      accountID,
		WatchingCount:  watchingCount,
		Items:          items,
		DataSourceNote: "user_opportunity_actions · latest per opportunity_id · WATCH only · trade_plan by stock_code",
	}, nil
}

// enrichWatchlistMeta derives display name/source from existing snapshot data (no new columns).
func enrichWatchlistMeta(scanBatchKey, stockCode string) (stockName, source string) {
	batch := strings.TrimSpace(scanBatchKey)
	code := NormalizeStockCode(stockCode)
	source = formatBatchSource(batch)

	if strings.HasPrefix(batch, "snap:") {
		idRaw := strings.TrimPrefix(batch, "snap:")
		id, err := strconv.ParseUint(idRaw, 10, 64)
		if err != nil || id == 0 {
			return "", source
		}
		var snap models.SignalScanSnapshot
		if err := db.Dao.First(&snap, uint(id)).Error; err != nil {
			return "", source
		}
		source = formatSnapshotSource(&snap)
		repo := data.NewSignalSnapshotRepo()
		for _, hit := range repo.ParseSnapshotHits(&snap) {
			if NormalizeStockCode(SecucodeToStockCode(hit.SECUCODE)) == code {
				return strings.TrimSpace(hit.SECURITY_NAME_ABBR), source
			}
		}
		return "", source
	}

	parts := strings.Split(batch, "|")
	if len(parts) == 3 {
		source = formatBatchSource(batch)
	}
	return "", source
}

func formatBatchSource(batch string) string {
	batch = strings.TrimSpace(batch)
	if batch == "" {
		return ""
	}
	if strings.HasPrefix(batch, "snap:") {
		return batch
	}
	parts := strings.Split(batch, "|")
	if len(parts) == 3 {
		return strings.TrimSpace(parts[0]) + " · " + strings.TrimSpace(parts[1]) + " · " + strings.TrimSpace(parts[2])
	}
	return batch
}

func formatSnapshotSource(snap *models.SignalScanSnapshot) string {
	if snap == nil {
		return ""
	}
	parts := []string{}
	if s := strings.TrimSpace(snap.TradeDate); s != "" {
		parts = append(parts, s)
	}
	if s := strings.TrimSpace(snap.Session); s != "" {
		parts = append(parts, s)
	}
	if s := strings.TrimSpace(snap.StrategyID); s != "" {
		parts = append(parts, s)
	}
	if len(parts) == 0 {
		return fmt.Sprintf("snap:%d", snap.ID)
	}
	return strings.Join(parts, " · ")
}

File: frontend/src/api/watchlist.ts
/**
 * Phase16.26-C2.3.2-A — GET /api/watchlist (read-only watching list + trade_plan_id).
 */

export type WatchlistItem = {
  opportunity_id?: string
  stock_code: string
  stock_name?: string
  scan_batch_key: string
  source?: string
  watch_time: string
  latest_action: string
  in_trade_plan?: boolean
  trade_plan_status?: string
  trade_plan_id?: number
}

export type WatchlistView = {
  account_id: number
  watching_count: number
  items: WatchlistItem[]
  data_source_note?: string
}

function mapItem(raw: Record<string, unknown>): WatchlistItem {
  const status = String(raw.trade_plan_status || raw.tradePlanStatus || 'NONE').toUpperCase() || 'NONE'
  const planId = Math.trunc(Number(raw.trade_plan_id ?? raw.tradePlanId) || 0)
  return {
    opportunity_id: raw.opportunity_id ? String(raw.opportunity_id) : undefined,
    stock_code: String(raw.stock_code || raw.stockCode || ''),
    stock_name: raw.stock_name ? String(raw.stock_name) : raw.stockName ? String(raw.stockName) : undefined,
    scan_batch_key: String(raw.scan_batch_key || raw.scanBatchKey || ''),
    source: raw.source ? String(raw.source) : undefined,
    watch_time: String(raw.watch_time || raw.watchTime || ''),
    latest_action: String(raw.latest_action || raw.latestAction || ''),
    in_trade_plan: Boolean(raw.in_trade_plan ?? raw.inTradePlan),
    trade_plan_status: status,
    trade_plan_id: planId > 0 ? planId : undefined,
  }
}

/** GET /api/watchlist */
export async function fetchWatchlist(opts: { accountId?: number; limit?: number } = {}): Promise<WatchlistView> {
  const params = new URLSearchParams()
  if (opts.accountId && opts.accountId > 0) params.set('account_id', String(opts.accountId))
  if (opts.limit && opts.limit > 0) params.set('limit', String(opts.limit))
  const qs = params.toString()
  const res = await fetch(`/api/watchlist${qs ? `?${qs}` : ''}`)
  const body = (await res.json()) as Record<string, unknown>
  if (!res.ok || !body?.ok) {
    throw new Error(String(body?.message || `HTTP ${res.status}`))
  }
  const wl = (body.watchlist || {}) as Record<string, unknown>
  const itemsRaw = Array.isArray(wl.items) ? wl.items : []
  return {
    account_id: Number(wl.account_id || wl.accountId) || 0,
    watching_count: Number(wl.watching_count || wl.watchingCount) || 0,
    items: itemsRaw.map((e) => mapItem(e as Record<string, unknown>)),
    data_source_note: wl.data_source_note ? String(wl.data_source_note) : undefined,
  }
}
